File: fushi/test/lookup/apostrophe_word_scan_bug2056_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:fushi/src/reader/reader_selection_scripts.dart';

/// BUG-2056：英文正文/字幕里点 `don’t` 的 don 只查到 "don"，点 t 只查到 "t"；
/// `it’s` / `John’s` / `we’ve` 这类**缩合形与所有格**整类匹配不到词条。
///
/// 根因：撇号是否词边界取决于**上下文**，不是字符本身，但 `scanDelimiters`
/// （四份实现逐字节相同）把 ASCII `'` 与排版撇号 `’`（U+2019，真实 EPUB 的主流
/// 写法）当成无条件扫描终点，于是 `selectFromPosition` 的前向扫描一撞上就 break。
///
/// 修复：加一条上下文判据 `isIntraWordApostrophe(text, index)`——撇号两侧都是
/// **空格分词类字母**时它是词内字符，前向扫描跨过去。字母集与
/// `native/fushidicts/fushidicts_src/scan/word_scan.cpp` 的
/// `is_space_delimited_letter` 逐区间对齐（全仓一个模型）。
///
/// **刻意不改词首回退。** 回退跨撇号会把法语/意大利语省音写法（l’homme、
/// dell’arte）的锚点从 homme 拖回 l’，反而查不到 homme；而前向跨过是纯增益——
/// C++ `scan_candidates` 会生成 `don’t` / `don’` / `don` 三级前缀，短词不会被
/// 挤掉。行为测试 ⑥⑦ 与本文件的源码守卫一起把这个取舍钉死。
///
/// 两层守护：
/// ① 行为级——用 node 真执行两份实现（浮窗/扩展的 selection.js + 阅读器注入脚本
///    `ReaderSelectionScripts.source()`）的 `selectFromPosition`，12 条断言覆盖
///    三种撇号、所有格+空格桥接叠加、引号语义、法语省音、跨节点、日文不回归。
///    无 node 时 skip。
/// ② 源码级——四份实现都必须在**终点判定之前**跨词内撇号，且词首回退保持原样。
void main() {
  const List<String> selectionCopies = <String>[
    'assets/popup/selection.js',
    'assets/browser_extension/vendor/selection.js',
    '../tools/browser-extension/vendor/selection.js',
  ];

  test(
    'BUG-2056: forward scan bridges an intra-word apostrophe so English '
    'contractions stay reachable (executes both selection engines via node)',
    () async {
      final String? nodeExe = _resolveNode();
      if (nodeExe == null) {
        markTestSkipped(
            'node not found on PATH; skipping JS behavior execution');
        return;
      }

      final File jsTest = File(
        'test/lookup/apostrophe_word_scan_bug2056_test.js',
      );
      expect(jsTest.existsSync(), isTrue,
          reason: 'behavior harness ${jsTest.path} must exist');

      // 阅读器注入脚本活在 Dart 的 raw string 里，node 读不到文件。把**真值**
      // （ReaderSelectionScripts.source()，与真机注入的是同一份字符串）落到临时
      // 文件再交给 harness，避免测试对着一份手抄副本自娱自乐。
      final Directory tmp =
          Directory.systemTemp.createTempSync('fushi_bug2056_');
      final File readerJs = File('${tmp.path}/reader_selection.js');
      readerJs.writeAsStringSync(ReaderSelectionScripts.source());

      try {
        final ProcessResult result = await Process.run(
          nodeExe,
          <String>[jsTest.path],
          workingDirectory: Directory.current.path,
          environment: <String, String>{
            'FUSHI_READER_SELECTION_JS': readerJs.path,
          },
        );

        expect(
          result.exitCode,
          0,
          reason: 'BUG-2056 intra-word apostrophe behavior test failed.\n'
              'stdout:\n${result.stdout}\nstderr:\n${result.stderr}',
        );
        final String stdout = result.stdout.toString();
        expect(stdout, contains('all assertions passed'),
            reason: 'behavior harness must reach its success marker');
        // 阅读器那一套必须真的跑到，否则这个测试只守住了一半实现。
        expect(stdout, isNot(contains('reader selection script not provided')),
            reason: '阅读器注入脚本必须被 harness 真执行（真值来自 source()）');
      } finally {
        tmp.deleteSync(recursive: true);
      }
    },
  );

  test(
      'intra-word apostrophe is bridged before the scan-stop test, and the '
      'word-start back-off stays untouched', () {
    final Map<String, String> sources = <String, String>{
      for (final String path in selectionCopies)
        path: File(path).readAsStringSync(),
      'reader_selection_scripts.dart (source())':
          ReaderSelectionScripts.source(),
    };

    sources.forEach((String label, String src) {
      // 顺序判据必须在**剥掉注释**的代码上算：两处修复的说明注释里都写着
      // isIntraWordApostrophe / isScanStop 这些标识符，裸 indexOf 会先命中注释，
      // 顺序断言就变成恒真（见 docs/agent 的顺序守卫纪律）。
      final String code = _stripLineComments(src);

      const String bridge =
          'if (this.isIntraWordApostrophe(content, scanOffset)) {';
      const String stop = 'if (this.isScanStop(char)) break;';

      expect(code.contains(bridge), isTrue, reason: '[$label] 前向扫描必须带词内撇号桥接');
      expect(code.contains(stop), isTrue,
          reason: '[$label] 终点判定必须仍在（BUG-1773 的锚点）');
      expect(code.indexOf(bridge) < code.indexOf(stop), isTrue,
          reason: '[$label] 撇号桥接必须**先于** isScanStop，否则撇号照旧 break');

      // 判据本身必须是「两侧都是空格分词类字母」，不是无条件跨过——无条件跨过会
      // 把 `‘hello’ world` 的收尾引号也吃掉、把日文里的引号粘进查询串。
      expect(code.contains('this.isSpaceDelimitedLetter(text[index - 1]) &&'),
          isTrue,
          reason: '[$label] 词内判据必须要求撇号**左**侧是空格分词类字母');
      expect(code.contains('this.isSpaceDelimitedLetter(text[index + 1]);'),
          isTrue,
          reason: '[$label] 词内判据必须要求撇号**右**侧是空格分词类字母');

      // 词首回退必须保持原样（整条 while 条件逐字不变）：把桥接加进回退会让
      // l’homme 点 homme 退回 l’，是净损失。行为测试 ⑥⑦ 守行为，这里守写法。
      expect(
          code.contains('while (startOffset > 0 && '
              '!this.isScanBoundary(hitContent[startOffset - 1])) {'),
          isTrue,
          reason: '[$label] 词首回退不得被改动（不得跨撇号）');
    });
  });
}

/// 去掉 `//` 行注释（保留换行，偏移仍然单调），供顺序判据使用。
String _stripLineComments(String src) {
  return src.split('\n').map((String line) {
    final int idx = line.indexOf('//');
    return idx < 0 ? line : line.substring(0, idx);
  }).join('\n');
}

/// Resolve a usable `node` executable, returning null when none is on PATH.
String? _resolveNode() {
  final List<String> candidates =
      Platform.isWindows ? <String>['node.exe', 'node'] : <String>['node'];
  for (final String name in candidates) {
    try {
      final ProcessResult probe = Process.runSync(name, <String>['--version']);
      if (probe.exitCode == 0) {
        return name;
      }
    } on ProcessException {
      // Not found; try next candidate.
    }
  }
  return null;
}

File: fushi/test/lookup/apostrophe_word_scan_bug2056_test.js
// BUG-2056 行为守卫：词内撇号必须被前向扫描跨过去，英语的缩合形/所有格才查得到。
//
// 症状：英文正文里点 "don’t" 的 don，喂给引擎的查询串是 "don"；点 t 是 "t"。
// en.json 词形还原表里 don't / it's / John's 这类整类匹配不到。真实 EPUB 几乎都用
// 排版撇号 U+2019，而它和 ASCII ' 一样在 scanDelimiters 里，前向扫描一撞上就 break。
//
// 根因修复：撇号是否词边界取决于**上下文**，不是字符本身。两侧都是空格分词类字母
// （字母集与 native/fushidicts/fushidicts_src/scan/word_scan.cpp 的
// is_space_delimited_letter 逐区间对齐）时它是词内字符，前向扫描跨过去。
//
// 刻意**不改词首回退**：回退跨撇号会把 l’homme / dell’arte 的锚点从 homme 拖回 l’，
// 反而查不到 homme。前向跨过是纯增益（scan_candidates 生成 don’t / don’ / don
// 三级前缀，短词不会被挤掉），回退跨过是零和的锚点搬家。⑥⑦ 两条把这个取舍钉死。
//
// 本 harness 用 node:vm 在最小 fake DOM 里真执行两份实现的 selectFromPosition
// （fake DOM 与 phrase_lookup_whitespace_bridge_bug1773_test.js 同构）：
//   ① assets/popup/selection.js（浮窗 / 浏览器扩展，三镜像逐字节 parity 另有测试守）
//   ② 阅读器注入脚本 ReaderSelectionScripts.source()——由 .dart wrapper 落到临时
//      文件，路径经环境变量 FUSHI_READER_SELECTION_JS 传入；没传就只测 ①。
//
// 运行：node fushi/test/lookup/apostrophe_word_scan_bug2056_test.js

const assert = require('node:assert');
const fs = require('node:fs');
const path = require('node:path');
const vm = require('node:vm');

// ---- 最小 fake DOM -------------------------------------------------------

function makeText(content, parent) {
  return { nodeType: 3, textContent: content, nodeValue: content, parentElement: parent };
}

function selectorTags(selector) {
  return selector
    .split(',')
    .map((s) => s.trim().toLowerCase())
    .filter((s) => s.length > 0 && !s.startsWith('.') && !s.startsWith('['));
}

function makeElement(tagName) {
  const el = {
    nodeType: 1,
    tagName,
    className: '',
    parentElement: null,
    childNodes: [],
    get textContent() {
      return this.childNodes.map((c) => c.textContent).join('');
    },
    closest(selector) {
      const tags = selectorTags(selector);
      let node = this;
      while (node && node.nodeType === 1) {
        if (tags.includes((node.tagName || '').toLowerCase())) return node;
        node = node.parentElement;
      }
      return null;
    },
  };
  return el;
}

function makeRange() {
  return {
    startContainer: null,
    startOffset: 0,
    endContainer: null,
    endOffset: 0,
    setStart(node, offset) {
      this.startContainer = node;
      this.startOffset = offset;
    },
    setEnd(node, offset) {
      this.endContainer = node;
      this.endOffset = offset;
    },
    collapse() {},
    getClientRects() {
      return [{ left: 0, right: 10, top: 0, bottom: 10, x: 0, y: 0, width: 10, height: 10 }];
    },
    getBoundingClientRect() {
      return { left: 0, right: 10, top: 0, bottom: 10, x: 0, y: 0, width: 10, height: 10 };
    },
  };
}

// TreeWalker：尊重 acceptNode 的 FILTER_REJECT（阅读器版据此跳过纯空白节点）。
function makeTreeWalker(root, filter) {
  const out = [];
  (function walk(node) {
    for (const child of node.childNodes || []) {
      if (child.nodeType === 3) {
        if (!filter || filter.acceptNode(child) === 1) out.push(child);
      } else if (child.nodeType === 1) {
        walk(child);
      }
    }
  })(root);
  return {
    currentNode: root,
    nextNode() {
      const from = out.indexOf(this.currentNode);
      const next = out[from + 1] || null;
      if (next) this.currentNode = next;
      return next;
    },
  };
}

/// 建一个 <p>，其中每段文字是一个 <span> 里的文本节点（segments 长度 > 1 时用来
/// 覆盖「跨文本节点续扫」）。返回 { sandbox, textNodes }。
function buildContext(src, segments) {
  const body = makeElement('body');
  const p = makeElement('p');
  p.parentElement = body;
  body.childNodes = [p];

  const textNodes = [];
  for (const seg of segments) {
    const span = makeElement('span');
    span.parentElement = p;
    const t = makeText(seg, span);
    span.childNodes = [t];
    p.childNodes.push(span);
    textNodes.push(t);
  }

  const document = {
    body,
    createRange: makeRange,
    createTreeWalker: (root, _whatToShow, filter) => makeTreeWalker(root, filter),
    caretPositionFromPoint: () => null,
    caretRangeFromPoint: () => null,
    elementFromPoint: () => null,
  };

  const window = {
    // 所有 span 都是行内盒 → crossesRenderBoundary 判「同一行连排」，跨节点续扫放行
    // （popup 版的 BUG-1645 路径；阅读器版没有这个函数，行为一致）。
    getComputedStyle: () => ({ display: 'inline', content: 'none' }),
    getSelection: () => ({ removeAllRanges() {} }),
    flutter_inappwebview: { callHandler: () => {} },
    scanNonJapaneseText: true,
  };
  const sandbox = {
    window,
    document,
    Node: { ELEMENT_NODE: 1, TEXT_NODE: 3 },
    NodeFilter: { SHOW_TEXT: 4, FILTER_ACCEPT: 1, FILTER_REJECT: 2 },
    CSS: undefined,
    console,
  };
  sandbox.globalThis = sandbox;
  vm.createContext(sandbox);
  vm.runInContext(src, sandbox);

  // 阅读器版的收尾走 fireTextSelected → buildSelectionPayload（依赖整套
  // window.fushiReader 归一化偏移）。本测试只关心扫描产出的查询串，stub 掉收尾。
  sandbox.window.fushiSelection.fireTextSelected = function () {
    return this.selection ? this.selection.text : null;
  };
  return { sandbox, textNodes };
}

/// 跑一次取词，返回喂给引擎的查询串（没选到返回 null）。
function scan(src, segments, nodeIndex, offset, maxLength) {
  const { sandbox, textNodes } = buildContext(src, segments);
  const sel = sandbox.window.fushiSelection;
  sel.selection = null;
  sel.selectFromPosition(textNodes[nodeIndex], offset, maxLength === undefined ? 24 : maxLength);
  return sel.selection ? sel.selection.text : null;
}

// ---- 断言 ----------------------------------------------------------------

// ---- 断言 ----------------------------------------------------------------

const RSQUO = '\u2019'; // ’ 排版撇号：真实 EPUB 里的主流写法
const MODAP = '\u02BC'; // ʼ MODIFIER LETTER APOSTROPHE

function runSuite(label, src) {
  // ① 根因回归：排版撇号必须被跨过，don’t 整体进查询串。
  //    修复前 offset 0/1/2 都只得到 "don"。
  for (const offset of [0, 1, 2]) {
    assert.strictEqual(
      scan(src, [`I don${RSQUO}t.`], 0, 2 + offset),
      `don${RSQUO}t`,
      `[${label}] 点 don 的第 ${offset} 位必须扫出跨撇号的 don’t（'.' 处终止）`,
    );
  }

  // ② ASCII 撇号同样处理（纯文本字幕/OCR 里常见）。
  assert.strictEqual(
    scan(src, ["don't."], 0, 0),
    "don't",
    `[${label}] ASCII 撇号也必须被跨过`,
  );

  // ③ U+02BC 同样处理。
  assert.strictEqual(
    scan(src, [`can${MODAP}t.`], 0, 0),
    `can${MODAP}t`,
    `[${label}] U+02BC 也必须被跨过`,
  );

  // ④ 所有格 + 空格桥接叠加：撇号跨过后，BUG-1773 的空格桥接照常接上下一个词。
  assert.strictEqual(
    scan(src, [`John${RSQUO}s book.`], 0, 0),
    `John${RSQUO}s book`,
    `[${label}] 所有格与空格桥接必须叠加生效`,
  );

  // ⑤ 引号语义不受影响：右侧是空白 → 撇号仍是扫描终点。
  assert.strictEqual(
    scan(src, [`${RSQUO}hello${RSQUO} world`], 0, 1),
    'hello',
    `[${label}] 收尾引号（右侧空白）必须仍是终点`,
  );

  // ⑥ **词首回退不得跨撇号**：法语省音 l’homme 点 homme 仍锚在 homme。
  //    若把桥接也加进回退，这里会退回 l’homme，反而查不到 homme。
  assert.strictEqual(
    scan(src, [`l${RSQUO}homme.`], 0, 3),
    'homme',
    `[${label}] 词首回退必须仍在撇号处停住（法/意省音不得被拖回前缀）`,
  );

  // ⑦ 同理，点 don’t 的 t 仍只得到 t——这是 ⑥ 的同一条规则，刻意保留。
  assert.strictEqual(
    scan(src, [`don${RSQUO}t.`], 0, 4),
    't',
    `[${label}] 点撇号右侧首字母仍只从该字母起扫（与 ⑥ 同一条规则）`,
  );

  // ⑧ 非空格分词脚本两侧的撇号仍是终点（日文正文里的 ’ 是引号，不是词内字符）。
  assert.strictEqual(
    scan(src, [`私は${RSQUO}そう${RSQUO}言った`], 0, 0),
    '私は',
    `[${label}] 日文两侧的撇号必须仍是终点`,
  );

  // ⑨ 跨文本节点时新节点开头的撇号不得桥接（与空白桥接同一条纪律）。
  assert.strictEqual(
    scan(src, ['don', `${RSQUO}t`], 0, 0),
    'don',
    `[${label}] 新文本节点开头的撇号不得被当成词内字符`,
  );

  // ⑩ 撇号后必须真有字母才跨：`rock 'n' roll` 的 ' 右侧是 n、左侧是空白 → 终点。
  assert.strictEqual(
    scan(src, ["rock 'n' roll"], 0, 0),
    'rock',
    `[${label}] 左侧非字母的撇号必须仍是终点`,
  );

  // ⑪ maxLength 仍是硬上限。
  assert.strictEqual(
    scan(src, [`don${RSQUO}t ask`], 0, 0, 4),
    `don${RSQUO}`,
    `[${label}] maxLength 必须仍然截断`,
  );

  // ⑫ 日文逐字扫描完全不受影响（回归护栏）。
  assert.strictEqual(
    scan(src, ['素晴らしい世界'], 0, 0),
    '素晴らしい世界',
    `[${label}] 日文扫描行为不变`,
  );
}

function run() {
  const popupSrc = fs.readFileSync(
    path.resolve(__dirname, '..', '..', 'assets', 'popup', 'selection.js'),
    'utf8',
  );
  runSuite('popup/extension', popupSrc);

  const readerJsPath = process.env.FUSHI_READER_SELECTION_JS;
  if (readerJsPath && fs.existsSync(readerJsPath)) {
    runSuite('reader', fs.readFileSync(readerJsPath, 'utf8'));
  } else {
    console.log('reader selection script not provided; skipped that suite');
  }

  console.log('all assertions passed');
}

run();
